### ISSUE TITLE
NullPointerException when parsing sentences in Arabic

### ISSUE THREAD

**Issue by mingdachen** [stanfordnlp/CoreNLP#966]
Hi,

I was trying to parse the sentences in Arabic from OpenSubtitles, but got the following error (pos tagging works fine):
```
Exception in thread "main" java.lang.NullPointerException
        at edu.stanford.nlp.parser.lexparser.ExhaustiveDependencyParser.getBestScore(ExhaustiveDependencyParser.java:607)
        at edu.stanford.nlp.parser.lexparser.ExhaustiveDependencyParser.hasParse(ExhaustiveDependencyParser.java:602)
        at edu.stanford.nlp.parser.lexparser.LexicalizedParserQuery.getBestParse(LexicalizedParserQuery.java:398)
        at edu.stanford.nlp.parser.lexparser.LexicalizedParserQuery.getBestParse(LexicalizedParserQuery.java:375)
        at edu.stanford.nlp.pipeline.ParserAnnotator.doOneSentence(ParserAnnotator.java:328)
        at edu.stanford.nlp.pipeline.ParserAnnotator.doOneSentence(ParserAnnotator.java:254)
        at edu.stanford.nlp.pipeline.SentenceAnnotator.annotate(SentenceAnnotator.java:102)
        at edu.stanford.nlp.pipeline.AnnotationPipeline.annotate(AnnotationPipeline.java:76)
        at edu.stanford.nlp.pipeline.StanfordCoreNLP.annotate(StanfordCoreNLP.java:637)
        at edu.stanford.nlp.pipeline.StanfordCoreNLP.annotate(StanfordCoreNLP.java:647)
        at edu.stanford.nlp.pipeline.StanfordCoreNLP.processFiles(StanfordCoreNLP.java:1226)
        at edu.stanford.nlp.pipeline.StanfordCoreNLP.processFiles(StanfordCoreNLP.java:1060)
        at edu.stanford.nlp.pipeline.StanfordCoreNLP.run(StanfordCoreNLP.java:1343)
        at edu.stanford.nlp.pipeline.StanfordCoreNLP.main(StanfordCoreNLP.java:1389)
```

Following is the command I used

```
java -cp stanford-corenlp-3.9.2.jar:stanford-arabic-corenlp-2018-10-05-models.jar \
     -Xmx2g edu.stanford.nlp.pipeline.StanfordCoreNLP \
     -tokenize.language ar \
     -segment.model edu/stanford/nlp/models/segmenter/arabic/arabic-segmenter-atb+bn+arztrain.ser.gz \
     -pos.model edu/stanford/nlp/models/pos-tagger/arabic/arabic.tagger \
     -parse.model edu/stanford/nlp/models/lexparser/arabicFactored.ser.gz \
     -threads 1 \
     -annotators tokenize,ssplit,parse \
     -ssplit.eolonly \
     -filelist intpu_file_list.txt -outputFormat text \
     -outputDirectory parsed_ar
```

Did I do anything wrong? Any help would be appreciated!

**Comment by AngledLuffa** [comment]
Can you isolate a few sentences which cause the problem?  Binary search on
the input text, for example

**Comment by mingdachen** [comment]
Hi @AngledLuffa thanks for the prompt reply! I managed to nail it down to [these 20 sentences](https://github.com/stanfordnlp/CoreNLP/files/3854042/failure_sentences.txt). It's weird that the problem seems go away if I made the number of sentences smaller.

**Comment by AngledLuffa** [comment]
I can't reproduce this using this file.  Are you sure the models and code
are both up to date?

What OS & java version are you using?

**Comment by mingdachen** [comment]
> I can't reproduce this using this file. Are you sure the models and code are both up to date? What OS & java version are you using?
> […](#)

I tried it on both Linux and Ubuntu. My java version is ```openjdk version "1.8.0_03-Ubuntu"``` and ```java version "1.8.0_101"``` 

I used models and code from the official website, which should be the same as what's available on Github since both are version 3.9.2, I think?

I can try models and code base on Github, but  I don't have any experience in doing this. Should I rebuild the whole thing and then download the model file?

**Comment by AngledLuffa** [comment]
Perhaps upgrading java would help?  I've tried it on

centos 6 w/ java 1.8.172
centos 8 w/ java 1.8.232
windows 10 w/ java 12.0.1

If you could double check that it was in those 20 sentences, that might
help.

John

**Comment by mingdachen** [comment]
Thanks for trying it. I think the problem may be related to the memory; When I split these 20 sentences into smaller files, I got no problem parsing it. And when I looked at the parsed files, it seems like the CoreNLP didn't treat line breaks as the boundaries of sentences, which I think could be the cause. But the same configuration seems to work fine for French and English (i.e. CoreNLP will use line breaks as the boundaries of sentences), do you know why it works differently for Arabic?

**Comment by AngledLuffa** [comment]
You have set it to split on eolonly, which is reasonable since it looks
like you have one sentence per line.  (I understand 0 Arabic but I see a .
! or ... at the end of each line).  Why it isn't splitting one sentence per
line seems kind of strange.  I'll have to take a look.

**Comment by AngledLuffa** [comment]
It turns out the ArabicSegmenterAnnotator wasn't implementing the eolonly
property at all.  The result is it was clumping all of the sentences into
one sentence.

Your very clear example was quite useful in terms of diagnosing and fixing
this, so thank you.

I have pushed a fix for the next version of corenlp.

Unfortunately, that will still be a few weeks off.  I don't think you'd be
able to use the current state of github, either, as the next release has
some partially finished changes in it.

However, what you can do (this will hopefully work) is recompile the source
included in stanford-corenlp-3.9.2-sources.jar after patching with these
two files:

https://github.com/stanfordnlp/CoreNLP/blob/master/src/edu/stanford/nlp/util/StringUtils.java

https://github.com/stanfordnlp/CoreNLP/blob/master/src/edu/stanford/nlp/pipeline/ArabicSegmenterAnnotator.java

It looks like the changes I just made haven't propagated from our internal
repo yet, so you need to wait a little while.  If you don't see
"splitLinesKeepNewlines" in StringUtils yet, the changes aren't available
yet.

**Comment by mingdachen** [comment]
Thanks for looking into this!
